File: backend/app/schemas/resume.py
from datetime import datetime
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, ConfigDict, Field


class ResumeCandidateInfo(BaseModel):
    name: Optional[str] = None
    phone: Optional[str] = None
    email: Optional[str] = None
    education: Optional[str] = None
    school: Optional[str] = None
    major: Optional[str] = None
    graduation_year: Optional[str] = None
    work_years: Optional[str] = None


class ResumeParseResult(BaseModel):
    candidate: ResumeCandidateInfo = Field(default_factory=ResumeCandidateInfo)
    skills: List[str] = Field(default_factory=list)
    project_experiences: List[Dict[str, Any]] = Field(default_factory=list)
    work_experiences: List[Dict[str, Any]] = Field(default_factory=list)
    certificates: List[str] = Field(default_factory=list)
    resume_summary: Optional[str] = None
    strengths: List[str] = Field(default_factory=list)
    risks: List[str] = Field(default_factory=list)
    suggested_interview_questions: List[str] = Field(default_factory=list)
    confidence: float = 0
    need_manual_review: bool = False
    warnings: List[str] = Field(default_factory=list)


class ResumeFileOut(BaseModel):
    id: int
    candidate_id: Optional[int] = None
    original_filename: str
    stored_filename: str
    file_path: str
    file_hash: str
    mime_type: Optional[str] = None
    file_size: Optional[int] = None
    parse_status: str
    raw_text: Optional[str] = None
    parsed_json: Optional[Dict[str, Any]] = None
    resume_summary: Optional[str] = None
    skill_tags: Optional[List[str]] = None
    confidence: float = 0
    error_message: Optional[str] = None
    uploaded_by: Optional[str] = None
    source: str
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class ResumeScreeningReportOut(BaseModel):
    id: int
    candidate_id: int
    resume_file_id: int
    position_name: Optional[str] = None
    job_requirements: Optional[str] = None
    match_score: float = 0
    match_level: Optional[str] = None
    match_reason: Optional[str] = None
    strengths: Optional[List[str]] = None
    risks: Optional[List[str]] = None
    missing_requirements: Optional[List[str]] = None
    suggested_interview_questions: Optional[List[str]] = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class ResumeUploadResponse(BaseModel):
    success: bool
    message: str
    candidate: Optional[Dict[str, Any]] = None
    resume_file: Optional[Dict[str, Any]] = None
    resume_parse_result: Optional[Dict[str, Any]] = None
    screening_report: Optional[Dict[str, Any]] = None
    warnings: List[str] = Field(default_factory=list)


class ResumeScreenRequest(BaseModel):
    position_name: str
    job_requirements: Optional[str] = None

File: frontend/src/types/resume.ts
export interface ResumeFile {
  id: number;
  candidate_id?: number | null;
  original_filename: string;
  parse_status: string;
  resume_summary?: string | null;
  skill_tags?: string[] | null;
  confidence: number;
  error_message?: string | null;
  uploaded_by?: string | null;
  source: string;
  created_at: string;
  updated_at: string;
  parsed_json?: Record<string, unknown> | null;
  raw_text?: string | null;
}

export interface ResumeScreeningReport {
  id: number;
  candidate_id: number;
  resume_file_id: number;
  position_name?: string | null;
  match_score: number;
  match_level?: string | null;
  match_reason?: string | null;
  strengths?: string[] | null;
  risks?: string[] | null;
  missing_requirements?: string[] | null;
  suggested_interview_questions?: string[] | null;
  created_at: string;
}

export interface ResumeParseCandidateInfo {
  name?: string | null;
  phone?: string | null;
  email?: string | null;
  education?: string | null;
  school?: string | null;
  major?: string | null;
  graduation_year?: string | null;
  work_years?: string | null;
}

export interface ResumeParseResult {
  candidate: ResumeParseCandidateInfo;
  skills: string[];
  project_experiences: Array<Record<string, unknown>>;
  work_experiences: Array<Record<string, unknown>>;
  certificates: string[];
  resume_summary?: string | null;
  strengths: string[];
  risks: string[];
  suggested_interview_questions: string[];
  confidence: number;
  need_manual_review: boolean;
  warnings: string[];
}

export interface ResumeUploadCandidate {
  id?: number;
  name?: string | null;
  phone?: string | null;
  email?: string | null;
  position_name?: string | null;
  education?: string | null;
  school?: string | null;
  major?: string | null;
  graduation_year?: string | null;
  work_years?: string | null;
  skills?: string[] | null;
  resume_summary?: string | null;
  resume_parse_status?: string | null;
  latest_resume_id?: number | null;
  match_score?: number | null;
  match_level?: string | null;
  stage?: string | null;
  source?: string | null;
}

export interface ResumeUploadResponse {
  success: boolean;
  message: string;
  candidate?: ResumeUploadCandidate | null;
  resume_file?: ResumeFile | null;
  resume_parse_result?: ResumeParseResult | null;
  screening_report?: ResumeScreeningReport | null;
  warnings: string[];
}

export interface ResumeScreenRequest {
  position_name: string;
  job_requirements?: string;
}

export interface CandidateResumeResponse {
  resume_file: ResumeFile;
  screening_report?: ResumeScreeningReport | null;
}
